modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -1,64 +1,156 @@
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(ordinal)).handleErrorWith { e =>
           logger
-            .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with ordinal=${ordinal.show}")
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds
+          .fold(pullGlobalSnapshotsFromRandomPeer)(pullGlobalSnapshotsFromMajority)
+          .handleErrorWith { e =>
+            logger
+              .warn(e)(s"Failure pulling global snapshots")
+              .as(
+                List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
+              )
+          }
+
+      private def pullGlobalSnapshotsFromMajority(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        lastGlobalSnapshotStorage.getOrdinal.flatMap {
+          _.fold {
+            pullLatestSnapshotWithMajorityHash(majorityPeerIds).map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+          } { lastStoredOrdinal =>
+            for {
+              msd <- getLatestMajoritySnapshotData(majorityPeerIds)
+              l0Peers <- globalL0ClusterStorage.getPeers
+              pulled <- pullVerifiedSnapshots(lastStoredOrdinal.next, msd, l0Peers)
+            } yield pulled.asRight[LatestSnapshotTuple]
+          }
+        }
+
+      private def pullLatestSnapshotWithMajorityHash(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] = {
+        type Agg = SortedSet[L0Peer]
+        type Result = LatestSnapshotTuple
+        for {
+          peers <- globalL0ClusterStorage.getPeers.map(nes => Random.shuffle(nes.toSortedSet))
+          majorityPeers <- getL0Peers(majorityPeerIds)
+          result <- peers.tailRecM[F, Result] { l0Peers =>
+            l0Peers.headOption.fold {
+              NoPeersWithMajorityHash.raiseError[F, Either[Agg, Result]]
+            } { l0Peer =>
+              pullLatestSnapshotFromPeer(l0Peer).flatMap {
+                case (snapshot, info) =>
+                  Applicative[F].ifF(verifyLatestSnapshot(snapshot, majorityPeers))(
+                    (snapshot, info).asRight[Agg],
+                    l0Peers.tail.asLeft[Result]
+                  )
+              }.handleErrorWith { err =>
+                logger
+                  .warn(err)(s"Error pulling latest snapshot from peer ${l0Peer.show}")
+                  .as(l0Peers.tail.asLeft[Result])
+              }
+            }
+          }
+        } yield result
+      }
+
+      private def verifyLatestSnapshot(
+        snapshot: Hashed[GlobalIncrementalSnapshot],
+        majorityPeers: NonEmptyList[L0Peer]
+      ): F[Boolean] =
+        Monad[F].ifM(
```

Comment:
We could do one last thing as a part of this verification step, which is check if the snapshot info's merkle root matches the one in the snapshot. See: `MerkleTreeValidatorSuite`